### PR TITLE
* NEW: BinSkim will now log warning `OneOrMoreSymbolServersProvidedInaccessible` if can not connect to the http symbol servers provided by `--sympath`

### PR DESCRIPTION
**Pull request by shaopeng-gh**
### Issue:

User provided:
`--sympath srv*https://username:password@artifacts.dev.azure.com/Org/_apis/Symbol/symsrv`
It does not work for some reason. It is running on server which is less easy to verify if issues unrelated to BinSkim itself:
1. network restrictions the url can not be reached
2. user name and password wrong

### Change:
Add the probe and when the user provided http symbol server can not be reached, we log a warning message.

### CODE REVIEW

File: src/BinSkim.Driver/DriverUtilities.cs [hunk 1]
```diff
@@ -45,5 +49,35 @@ private static ConventionBuilder GetConventions<T>()
...
+
+            string[] symbolServers = Array.FindAll(symbolPathString.Split('*'), path =>
+            path.StartsWith("http://", StringComparison.OrdinalIgnoreCase) ||
+            path.StartsWith("https://", StringComparison.OrdinalIgnoreCase));
+
+            foreach (string symbolServer in symbolServers)
+            {
+                try
+                {
+                    var httpClient = new HttpClientWrapper();
```
Comment (suvamM) [inline src/BinSkim.Driver/DriverUtilities.cs:66]: @shaopeng-gh Wanted to cross-check that this invocation does not return a new `HttpClient` in each iteration. If so, we should create a single client and re-use that.

File: src/BinSkim.Driver/DriverUtilities.cs [hunk 2]
```diff
@@ -45,5 +49,35 @@ private static ConventionBuilder GetConventions<T>()
...
+                    HttpResponseMessage httpResponseMessage = httpClient.GetAsync(symbolServer).GetAwaiter().GetResult();
```
Comment (suvamM) [inline src/BinSkim.Driver/DriverUtilities.cs:67]: We are doing to an HTTP GET here: what is the authentication model? Binskim mostly runs in the context of Guardian---will that have sufficient privileges? Or will the majority of these checks result in an inaccessible error?

File: src/Test.UnitTests.BinSkim.Driver/DriverUtilitiesTests.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+using Microsoft.CodeAnalysis.IL;
+
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.BinSkim.Rules
+{
+    public class DriverUtilitiesTests
+    {
+        [Fact]
+        public void GetInaccessibleSymbolPathsTests()
```
Comment (suvamM) [inline src/Test.UnitTests.BinSkim.Driver/DriverUtilitiesTests.cs:17]: @shaopeng-gh This isn't really a unit test right? Will these result in actual network calls? If yes, then you need to mock the `HttpClient` to make this a unit test. We need separate integration tests to ensure that network calls work properly, but again the authentication model is unclear to me.

File: src/BinSkim.Driver/DriverUtilities.cs [hunk 4]
```diff
@@ -45,5 +49,35 @@
 
             return conventions;
         }
+
+        public static IList<string> GetInaccessibleSymbolPaths(string symbolPathString)
```
Comment (michaelcfanning) [inline src/BinSkim.Driver/DriverUtilities.cs:53]: >public [](http://example.com/codeflow?start=8&length=6)

make this internal

File: src/BinSkim.Driver/DriverUtilities.cs [hunk 5]
```diff
@@ -45,5 +49,35 @@
...
+        {
+            IList<string> inaccessibleList = new List<string>();
+            if (string.IsNullOrWhiteSpace(symbolPathString)) { return inaccessibleList; }
+
+            string[] symbolServers = Array.FindAll(symbolPathString.Split('*'), path =>
+            path.StartsWith("http://", StringComparison.OrdinalIgnoreCase) ||
```
Comment (michaelcfanning) [inline src/BinSkim.Driver/DriverUtilities.cs:59]: >path [](http://example.com/codeflow?start=12&length=4)

this should be indented.

File: src/BinSkim.Driver/MultithreadedAnalyzeCommand.cs [hunk 6]
```diff
@@ -69,6 +69,24 @@ public override BinaryAnalyzerContext InitializeGlobalContextFromOptions(Analyze
...
+                // These symbol servers provided are inaccessible: {0}.
+                context.Logger.LogConfigurationNotification(
+                    Errors.CreateNotification(
+                        context.CurrentTarget?.Uri,
+                        "OneOrMoreSymbolServersProvidedInaccessible",
+                        ruleId: null,
+                        FailureLevel.Warning,
+                        exception: null,
+                        persistExceptionStack: false,
+                        messageFormat: RuleResources.OneOrMoreSymbolServersProvidedInaccessible,
```
Comment (michaelcfanning) [inline src/BinSkim.Driver/MultithreadedAnalyzeCommand.cs:86]: >OneOrMoreSymbolServersProvidedInaccessible [](http://example.com/codeflow?start=53&length=42)

OneOrMoreSymbolServersAreNotAccessible.

When you create user-facing strings, please provide a real example as surfaced by the tool.

File: src/BinSkim.Driver/MultithreadedAnalyzeCommand.cs [hunk 7]
```diff
@@ -69,6 +69,24 @@ public override BinaryAnalyzerContext InitializeGlobalContextFromOptions(Analyze
 
             // Update context object based on command-line parameters.
             context.SymbolPath = options.SymbolsPath;
+
+            IList<string> inaccessibleSymbolPaths = DriverUtilities.GetInaccessibleSymbolPaths(context.SymbolPath);
+
+            if (inaccessibleSymbolPaths.Count > 0)
+            {
+                // These symbol servers provided are inaccessible: {0}.
```
Comment (michaelcfanning) [inline src/BinSkim.Driver/MultithreadedAnalyzeCommand.cs:77]: > // These symbol servers provided are inaccessible: {0}. [](http://example.com/codeflow?start=15&length=56)

The following symbol servers were specified via the --sympath argument but can't be reached:

Comment (michaelcfanning) [reply src/BinSkim.Driver/MultithreadedAnalyzeCommand.cs:77]: One or more servers specified in the --sympath argument could not be reached ([https://symweb](https://symweb), [https://blah](https://blah)).

File: src/Test.UnitTests.BinSkim.Driver/DriverUtilitiesTests.cs [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+        {
+            DriverUtilities.GetInaccessibleSymbolPaths(null)
+                .Should().BeEquivalentTo(Array.Empty<string>());
```
Comment (michaelcfanning) [inline src/Test.UnitTests.BinSkim.Driver/DriverUtilitiesTests.cs:20]: >()); [](http://example.com/codeflow?start=60&length=4)

Use xunit scopes here?

Test the output of your test  failures please!

File: src/Test.UnitTests.BinSkim.Driver/DriverUtilitiesTests.cs [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+            DriverUtilities.GetInaccessibleSymbolPaths("")
+                .Should().BeEquivalentTo(Array.Empty<string>());
+            DriverUtilities.GetInaccessibleSymbolPaths("SRV*http://msdl.microsoft.com/download/symbols")
+                .Should().BeEquivalentTo(Array.Empty<string>());
+            DriverUtilities.GetInaccessibleSymbolPaths("SRV*c:\\symbols*http://msdl.microsoft.com/download/symbols")
+                .Should().BeEquivalentTo(Array.Empty<string>());
+            DriverUtilities.GetInaccessibleSymbolPaths("CACHE*c:\\symbols;SRV*http://notexists/SymStore*http://notexists2/SymStore")
```
Comment (michaelcfanning) [inline src/Test.UnitTests.BinSkim.Driver/DriverUtilitiesTests.cs:27]: >notexists [](http://example.com/codeflow?start=85&length=9)

[https://does.notexist.example.com](https://does.notexist.example.com)

**Pull request closed without merging**